tools/envoy/fetch_sources.sh

```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
+# This script fetches Envoy source code to $SOURCE_DIR
+#
+# Requires:
+# - $SOURCE_DIR, a directory where sources will be placed
+#
+# Optional:
+# - $ENVOY_TAG, git tag to reference specific revision
+# - $ENVOY_COMMIT_HASH, hash of the git commit. If specified, then $ENVOY_TAG will be ignored
+#
+# at least one of $ENVOY_TAG or $ENVOY_COMMIT_HASH should be specified
+set -e
+
+function msg_red() {
+  builtin echo -en "\033[1;31m" >&2
+  echo "$@" >&2
+  builtin echo -en "\033[0m" >&2
+}
+
+function msg_err() {
+  msg_red "$@"
+  exit 1
+}
+
+[[ -z "${SOURCE_DIR}" ]] && msg_err "Error: SOURCE_DIR is not specified"
+[[ -z "${ENVOY_TAG}" ]] && [[ -z "${ENVOY_COMMIT_HASH}" ]] && msg_err "Error: either ENVOY_TAG or ENVOY_COMMIT_HASH should be specified"
```

Comment:
That will work for `fetch_sources.sh`, but `fetch.sh` is using different steps depending on what was provided (tag or commit_hash). Technically if we'll name envoy binaries like `envoy-${TAG_OR_HASH}-${PLATFORM}` then it will be possible, but I think `envoy-1.20.1-dev-b16d390f-darwin` is more readable than `envoy-b16d390f-darwin`, what do you think?

Reply:
You won't download envoy twice in `envoy/fetch.sh`, there is `exit 0` in every `if` branch:
```bash
if [[ -n "${ENVOY_COMMIT_HASH}" ]]; then
  ENVOY_SHORT_HASH=${ENVOY_COMMIT_HASH:0:8}

  BINARY_NAME=$(curl --silent https://download.konghq.com/mesh-alpine/ \
    | { grep "${ENVOY_SHORT_HASH}" || true; } \
    | { grep "${ENVOY_DISTRO}" || true; } \
    | sed -e 's#.*<li><a href=".*">\(.*\)</a></li>#\1#')

  [[ -z "${BINARY_NAME}" ]] && msg_err "failed to resolve binary name by ENVOY_COMMIT_HASH=${ENVOY_COMMIT_HASH}"

  download_envoy "${BINARY_NAME}"
  exit 0
fi
```
Both `envoy/fetch.sh` and `envoy/fetch_sources.sh` prefer commit over the tag.